Basic texture handling

Quaker762 opened the pull request:
Implemented some basic texture handling, this relies on holding a reference to sh3_arc, so definitely needs to be refactored.

Quaker762 committed:
Basic texture handling

Quaker762 commented:
I'm going to commit this tomorrow because I need to go to bed. If you want to, you can do up all of the doxygen stuff, or leave it to me for tomorrow.

z33ky commented on include/SH3/graphics/texture.hpp:
```diff
@@ -1,56 +1,89 @@
+ *
+ *  Functions and structures to load a texture from an arc section
+ *
+ *  \copyright 2016  Palm Studios
```
We should choose either consistently `@tag` or `\tag`. I made the same error in my doxygen branch...

Quaker762 replied:
Okay, I like the `@tag` better, so let's stick to that.

z33ky commented on include/SH3/graphics/texture.hpp:
```diff
@@ -1,56 +1,89 @@
+    typedef struct
+    {
+        std::uint32_t batchHeaderMarker;          // This should be 0xFFFFFFFF to mark a header chunk
+        std::uint32_t unused1;                    // There are a lot of unused DWORDs, I assume to align everything nicely
```
Perhaps it would be easier to understand byte-sized, instead of 4-byte sized, unused fields.

z33ky commented on include/SH3/graphics/texture.hpp:
```diff
@@ -1,56 +1,89 @@
+     */
+    typedef struct
+    {
+        std::uint32_t batchHeaderMarker;          // This should be 0xFFFFFFFF to mark a header chunk
```
Use `/**< This should be 0xFFFFFFFF to mark a header chunk */` to document this field in Doxygen.

z33ky replied:
This is not 100% fixed. You're missing the `<`. Without it the documentation will be attached to the field following the comment, instead of the field preceding it.

z33ky commented on include/SH3/graphics/texture.hpp:
```diff
@@ -1,56 +1,89 @@
+namespace sh3_graphics
+{
+    /**
+     * Full texture header (both batch and individual texture).
```
@Add an empty comment line between `Full texture header` and the rest, because the first line is treated as a brief description by Doxygen.

Essentially, I think it should look like this:
```
   /** 
    *  Full texture header
    *  
    *  Both batch and individual texture.
    */
```

Also, I used two spaces after the asterisk, that aligns it better to the multi-line comment start IMO.

z33ky commented on include/SH3/graphics/texture.hpp:
```diff
@@ -1,56 +1,89 @@
+    class texture
+    {
+    public:
+    enum class result
```
Deliberately not indented?

Quaker762 replied:
I don't quite understand what you mean?

z33ky replied:
Stuff in `class textures` are indented two levels (8 spaces), except for the `enum` definition, which is only indented one level (4 spaces).

z33ky commented on include/SH3/graphics/texture.hpp:
```diff
@@ -1,56 +1,89 @@
+    public:
+    enum class result
+    {
+        INVALID_CHUNK,
```
My `enum class`es used CamelCase. We should decide on a consistent style.

Quaker762 replied:
Okay, I usually use Upper case with Underscores for anything constant (defines and enums), but I'll leave it up to you to decide which is better.

z33ky replied:
There is no better, it's just about consistency. I've reserved UPPER_CASE for preprocessor macros in my C++ stuff and CamelCase for everything class-related, that is class, struct & enum names, member functions (`static`) and methods, member and class/struct (`static`) variables and enum values. Other stuff is camelCase, that is class/struct-less functions, parameters, global and local variables. This has the nice side-effect that you can easily have a parameter, say in the constructor, with the same name as a member variable, because it is capitalized differently. It is a rather uncommon convention though.

z33ky commented on include/SH3/graphics/texture.hpp:
```diff
@@ -1,56 +1,89 @@
-        all of the file types, including all of the textures, meshes and even the motion capture
-        skeletal animation that Konami captured. Thanks mate!
+        texture(sh3_arc& arcref){arc = arcref;}
+        texture(std::string& filename, sh3_arc& arcref){arc = arcref; Load(filename);}
```
`filename` can be `const`.

z33ky commented on source/SH3/graphics/texture.cpp:
```diff
@@ -0,0 +1,36 @@
+{
+    if(arc.LoadFile(filename, pixbuff) == ARC_FILE_NOT_FOUND)
+    {
+        die("sh3_graphics::texture::Load( ): Unable to find file %s!", filename.c_str());
```
Should this not simply be encoded in `texture::result` as another error condition?

z33ky reviewed (changes requested):
To avoid storing a `sh3_arc` in `texture` you can simply pass it as another parameter to `Load` besides the filename.

z33ky commented on include/SH3/graphics/texture.hpp:
```diff
@@ -1,56 +1,89 @@
-Author:
-        Jesse Buhagiar
+    /**
+     * \class texture
```
You don't need to manually specify this.

z33ky commented on include/SH3/graphics/texture.hpp:
```diff
@@ -1,56 +1,89 @@
+    /**
+     * \class texture
+     *
+     * \brief Describes a logical texture and files to
```
At least in the `Doxyfile` of my doxygen branch, I've enabled autobrief to take the first line as brief description, which would eliminate the need for the brief-tag.

z33ky commented on source/SH3/graphics/texture.cpp:
```diff
@@ -0,0 +1,36 @@
+
+using namespace sh3_graphics;
+
+/**
```
Documentation should be in the header, because you're documenting the interface.

z33ky commented on source/SH3/graphics/texture.cpp:
```diff
@@ -0,0 +1,36 @@
+ * \return \c result indicating whether the read failed or not
+ *
+ */
+texture::result sh3_graphics::texture::Load(const std::string& filename)
```
Missing documentation for filename parameter.

Quaker762 committed:
Fixed Up texture documentation

Made a few fixes as recommended by Alexander. All doxy comments now use
@tag instead of \tag notation. Also changed sh3_texture_header uint32_t
to uint8_t[] arrays, as they are somewhat easier to read (though they
are all zero values)

z33ky commented on include/SH3/graphics/texture.hpp:
```diff
@@ -70,20 +72,28 @@ namespace sh3_graphics
+     *
+     * @param filename - Virtual file path to file.
+     *
+     * @return \c result indicating whether the read failed or not
```
Use `@ref` instead of `\c` so that Doxygen links the "result" to the type.

z33ky commented on include/SH3/graphics/texture.hpp:
```diff
@@ -70,20 +72,28 @@ namespace sh3_graphics
+     * @return \c result indicating whether the read failed or not
+     *
+     */
+        result Load(const std::string& filename, sh3_arc& arc);
```
`arc` is not documented.

z33ky commented on include/SH3/graphics/texture.hpp:
```diff
@@ -70,20 +72,28 @@ namespace sh3_graphics
-        texture(sh3_arc& arcref){arc = arcref;}
-        texture(std::string& filename, sh3_arc& arcref){arc = arcref; Load(filename);}
+        texture(){}
+        texture(std::string& filename, sh3_arc& arc){Load(filename, arc);}
```
`filename` can be `const`.

z33ky reviewed (changes requested):
Just two additional things:

* `result` should be `load_result` to it's clear that the `result` belongs to the `Load` method.
* What's with the empty `@note`s?

Then just squash it again and I won't be bikeshedding any longer ;)

Quaker762 committed:
Fixed bikeshedding changes [laughs]
